include/range/v3/action/insert.hpp

```diff
@@ -78,22 +78,46 @@ namespace ranges
 
             template<typename Cont, typename P, typename I, typename S,
                 typename C = common_iterator<I, S>,
-                CONCEPT_REQUIRES_(LvalueContainerLike<Cont>() && Iterator<P>() && IteratorRange<I, S>())>
+                CONCEPT_REQUIRES_(LvalueContainerLike<Cont>() && Iterator<P>() && IteratorRange<I, S>() &&
+                                  !(RandomAccessReservable<Cont>() && SizedIteratorRange<I, S>()))>
             auto insert(Cont && cont, P p, I i, S j) ->
                 decltype(unwrap_reference(cont).insert(p, C{i}, C{j}))
             {
                 return unwrap_reference(cont).insert(p, C{i}, C{j});
             }
 
+            template<typename Cont, typename P, typename I, typename S,
+                typename C = common_iterator<I, S>,
+                CONCEPT_REQUIRES_(RandomAccessReservable<Cont>() && Iterator<P>() && SizedIteratorRange<I, S>())>
+            auto insert(Cont && cont, P p, I i, S j) ->
+                decltype(unwrap_reference(cont).insert(begin(unwrap_reference(cont)), C{i}, C{j}))
+            {
+                auto const index = p - unwrap_reference(cont).begin();
+                unwrap_reference(cont).reserve(unwrap_reference(cont).size() + (j - i));
+                return unwrap_reference(cont).insert(begin(unwrap_reference(cont)) + index, C{i}, C{j});
+            }
+
             template<typename Cont, typename I, typename Rng,
                 typename C = range_common_iterator_t<Rng>,
-                CONCEPT_REQUIRES_(LvalueContainerLike<Cont>() && Iterator<I>() && Range<Rng>())>
+                CONCEPT_REQUIRES_(LvalueContainerLike<Cont>() && Iterator<I>() && Range<Rng>() &&
+                                  !(RandomAccessReservable<Cont>() && SizedRange<Rng>()))>
             auto insert(Cont && cont, I p, Rng && rng) ->
                 decltype(unwrap_reference(cont).insert(p, C{begin(rng)}, C{end(rng)}))
             {
                 return unwrap_reference(cont).insert(p, C{begin(rng)}, C{end(rng)});
             }
 
+            template<typename Cont, typename I, typename Rng,
+                typename C = range_common_iterator_t<Rng>,
+                CONCEPT_REQUIRES_(RandomAccessReservable<Cont>() && SizedRange<Rng>())>
+            auto insert(Cont && cont, I p, Rng && rng) ->
+                decltype(unwrap_reference(cont).insert(begin(unwrap_reference(cont)), C{begin(rng)}, C{end(rng)}))
+            {
+                auto const index = p - begin(unwrap_reference(cont));
+                unwrap_reference(cont).reserve(unwrap_reference(cont).size() + size(rng));
+                return unwrap_reference(cont).insert(begin(unwrap_reference(cont)) + index, C{begin(rng)}, C{end(rng)});
+            }
+
```

Comment:
I try to avoid negating concepts in requires clauses. Feels wrong stylistically. I also prefer keeping the count of public member functions down. This can be cleaned up by keeping the existing overloads and moving the implementation into some private members, dispatching to the correct one depending on the value of `RandomAccessReservable<Cont>() && SizedRange<Rng>()`.